File: pt5/1116.print-zero-even-odd.java
import java.util.concurrent.Semaphore;
import java.util.function.IntConsumer;

class ZeroEvenOdd {
    private final int n;
    private final Semaphore zeroTurn = new Semaphore(1);
    private final Semaphore oddTurn = new Semaphore(0);
    private final Semaphore evenTurn = new Semaphore(0);

    public ZeroEvenOdd(int n) {
        this.n = n;
    }

    public void zero(IntConsumer printNumber) throws InterruptedException {
        for (int value = 1; value <= n; value++) {
            zeroTurn.acquire();
            printNumber.accept(0);
            if ((value & 1) == 1) {
                oddTurn.release();
            } else {
                evenTurn.release();
            }
        }
    }

    public void even(IntConsumer printNumber) throws InterruptedException {
        for (int value = 2; value <= n; value += 2) {
            evenTurn.acquire();
            printNumber.accept(value);
            zeroTurn.release();
        }
    }

    public void odd(IntConsumer printNumber) throws InterruptedException {
        for (int value = 1; value <= n; value += 2) {
            oddTurn.acquire();
            printNumber.accept(value);
            zeroTurn.release();
        }
    }
}

File: pt5/1116.print-zero-even-odd.go
package main

type ZeroEvenOdd struct {
	n        int
	zeroTurn chan struct{}
	oddTurn  chan struct{}
	evenTurn chan struct{}
}

func Constructor(n int) ZeroEvenOdd {
	zeroTurn := make(chan struct{}, 1)
	oddTurn := make(chan struct{}, 1)
	evenTurn := make(chan struct{}, 1)
	zeroTurn <- struct{}{}
	return ZeroEvenOdd{n: n, zeroTurn: zeroTurn, oddTurn: oddTurn, evenTurn: evenTurn}
}

func (z *ZeroEvenOdd) zero(printNumber func(int)) {
	for value := 1; value <= z.n; value++ {
		<-z.zeroTurn
		printNumber(0)
		if value%2 == 1 {
			z.oddTurn <- struct{}{}
		} else {
			z.evenTurn <- struct{}{}
		}
	}
}

func (z *ZeroEvenOdd) even(printNumber func(int)) {
	for value := 2; value <= z.n; value += 2 {
		<-z.evenTurn
		printNumber(value)
		z.zeroTurn <- struct{}{}
	}
}

func (z *ZeroEvenOdd) odd(printNumber func(int)) {
	for value := 1; value <= z.n; value += 2 {
		<-z.oddTurn
		printNumber(value)
		z.zeroTurn <- struct{}{}
	}
}

func (z *ZeroEvenOdd) Zero(printNumber func(int)) {
	z.zero(printNumber)
}

func (z *ZeroEvenOdd) Even(printNumber func(int)) {
	z.even(printNumber)
}

func (z *ZeroEvenOdd) Odd(printNumber func(int)) {
	z.odd(printNumber)
}
